### ISSUE TITLE
Collect Data in a specific period (Since - Until)

### ISSUE THREAD

**Issue by pvita** [instaloader/instaloader#121]
Hi there,
is it possible to collect pictures (and metadata) only from a specific period? 
For example all the pictures of the last year with a specific hashtag.

P.s. is it possible to collect just metadata (+comments; geotags etc) without downloading all the pictures?

Best
P

**Comment by Thammus** [comment]
Hi,
you can play around with `dropwhile` and `takewhile` from [itertools](https://docs.python.org/3/library/itertools.html) like in the following snippet:
```python
from datetime import datetime
from itertools import dropwhile, takewhile

import instaloader

L = instaloader.Instaloader()

posts = L.get_hashtag_posts('milfgarden')
# or
# posts = instaloader.Profile.from_username(L.context, PROFILE).get_posts()

SINCE = datetime(2015, 5, 1)
UNTIL = datetime(2015, 3, 1)

for post in takewhile(lambda p: p.date > UNTIL, dropwhile(lambda p: p.date > SINCE, posts)):
    print(post.date)
    L.download_post(post, '#milfgarden')
```

**Comment by stagenam3** [comment]
Hi! I tried out that feature (copy pasted the code, changed the hashtag and dates, and ran the script)  and I get nothing, i.e. I do not get any error message and the script seems to have ran but I also do not get any results.

**Comment by suecys** [comment]
Hi, I have the same error as stagenam3, I get nothing when I change the dates and the hashtag. If I change the hashtag only, still nothing. What can I do to get results?

**Comment by oktayekici** [comment]
Same problem here. It does nothing after you run the code like the code has finished conducting.

**Comment by rohimari** [comment]
I get a problem, 429 Too Many Requests, how I can resolve it?

**Comment by rohimari** [comment]
> Hi,
> you can play around with `dropwhile` and `takewhile` from [itertools](https://docs.python.org/3/library/itertools.html) like in the following snippet:
> 
> ```python
> from datetime import datetime
> from itertools import dropwhile, takewhile
> 
> import instaloader
> 
> L = instaloader.Instaloader()
> 
> posts = L.get_hashtag_posts('milfgarden')
> # or
> # posts = instaloader.Profile.from_username(L.context, PROFILE).get_posts()
> 
> SINCE = datetime(2015, 5, 1)
> UNTIL = datetime(2015, 3, 1)
> 
> for post in takewhile(lambda p: p.date > UNTIL, dropwhile(lambda p: p.date > SINCE, posts)):
>     print(post.date)
>     L.download_post(post, '#milfgarden')
> ```

I can get a result from that code, but when I changed a hashtag and date, I get an error, 429 Too Many Requests.

**Comment by Alexhazl** [comment]
I have the same problem, I do not get the requests specified in the code and, when the process finishes, a KeyError: 'query_hash' 

What is happening?

**Comment by synchronizing** [comment]
This is the correct way about this:

```python
from datetime import datetime
import instaloader

def gen(start, end, posts):
    for post in posts:
        if post.date > start and post.date > end:
            pass
        elif post.date >= start and post.date <= end:
            yield post
        elif post.date <= start:
            break

L = instaloader.Instaloader()
profile = instaloader.Profile.from_username(L.context, "instagram")
posts = profile.get_posts()

for post in gen(start=datetime(2020, 11, 27), end=datetime(2020, 11, 30), posts=posts):
    print(post)
```

Note, however, that `end` is **not** inclusive, and `start` **is** inclusive. This because `end` (as in the example) is technically considered `datetime.datetime(2020, 11, 30, 0, 0)`, or 11-30-2020 at 00:00:00, which means a post at 11-30-2020 15:00:00 would not be included.

**Comment by synchronizing** [comment]
Added this feature for use out-of-the-box in #898.

**Comment by mahi17061992** [comment]
The below code works for me to get the posts from the last 7 days.

```
import instaloader
import datetime
import time
import csv
from pathlib import Path
from datetime import datetime, timedelta
from itertools import dropwhile, takewhile


def getposts_sevendays():
    loginuser = input('Enter Login User Name: ')
    loginpwd = input('Enter Login password: ')

    instanter = instaloader.Instaloader()
    # User need to login to his own account before accessing other followed/Public profiles.
    instanter.login(loginuser, loginpwd)

    fpath = Path('insta.csv').absolute()

    username = []
    with open(fpath, newline='') as f:
        for i in csv.reader(f):
            username.append(i[0])

    for user in username:

        posts = instaloader.Profile.from_username(instanter.context, user).get_posts()

        since = datetime.today()

        until = since - timedelta(days=7)

        for post in takewhile(lambda p: p.date > until, dropwhile(lambda p: p.date > since, posts)):
            print(post.date)
            instanter.download_post(post, user)

        time.sleep(60)
```

**Comment by piriyaraj** [comment]
```
from datetime import datetime
from itertools import dropwhile, takewhile

import instaloader

L = instaloader.Instaloader()

posts = instaloader.Profile.from_username(L.context, "instagram").get_posts()

SINCE = datetime(2015, 5, 1)
UNTIL = datetime(2015, 3, 1)

for post in takewhile(lambda p: p.date > UNTIL, dropwhile(lambda p: p.date > SINCE, posts)):
    print(post.date)
    L.download_post(post, "Instagram")
```


when running this code I get an error that is

**TypeError: can't compare offset-naive and offset-aware datetimes**

**at line** 
for post in takewhile(lambda p: p.date > UNTIL, dropwhile(lambda p: p.date > SINCE, posts)):

**Comment by aandergr** [comment]
> when running this code I get an error that is
> 
> **TypeError: can't compare offset-naive and offset-aware datetimes**
> 
> **at line** for post in takewhile(lambda p: p.date > UNTIL, dropwhile(lambda p: p.date > SINCE, posts)):

Thanks! This has been fixed with version 4.8.4.

**Comment by Demi-O** [comment]
Just going to add to this for those that might run into the problem I did. The solution here does not always work when trying to get posts from the last X days, so between `now` and `sometime in the past`, this is because pinned instagram posts come first in the list of posts regardless of when they were posted. So if the first pinned post is further in the past compared to your `UNTIL` date i.e `p.date > UNTIL` evaluates to `False`, then no other posts are returned even if they are more recent.

I modified @synchronizing's generator function above to what's below and it worked for me, unfortunately the `is_pinned` property for a pinned post I was testing with was returning `False` (the documentation actually warns about this, https://instaloader.github.io/module/structures.html#instaloader.Post.is_pinned) so I have to always check 3 extra posts to make sure that I am not skipping posts that are in the time range I want. 
```
MAX_NUMBER_OF_PINNED_POSTS = 3

def post_generator(start_time, end_time, posts):
    pinned_posts_counter = 0
    for post in posts:
        if post.date_utc > start_time and post.date_utc > end_time:
            pass
        elif post.date_utc >= start_time and post.date_utc <= end_time:
            yield post
        elif post.date_utc < start_time:
            if pinned_posts_counter < MAX_NUMBER_OF_PINNED_POSTS:
                pinned_posts_counter += 1
                continue
            else:
                break
```
I was wrongly assuming (probably most people are as well) that the posts are always sorted in chronological order starting with the most recent ones, found out the hard way that that's not always the case haha, but anyway hope this helps someone.
